### PR TITLE
Feature/darknet

### PR DESCRIPTION
**Pull request by junjihashimoto**

### CODE REVIEW

File: hasktorch/src/Torch/NN.hs [hunk 1]
```diff
@@ -48,11 +51,36 @@ instance Parameterized Parameter where
   flattenParameters = pure
   replaceOwnParameters _ = nextParameter
 
+instance Parameterized a => Parameterized [a] where
```
Comment (tycho01) [inline hasktorch/src/Torch/NN.hs:54]: for this instance maybe you can also use this [generalized version](https://github.com/hasktorch/hasktorch/pull/346/files#diff-1a12fff35c483c6404f1f2749298c9d1) (sorry I hadn't gotten around to fixing the other one there yet, but this one seems good).

File: hasktorch/src/Torch/NN.hs [hunk 2]
```diff
@@ -48,11 +51,36 @@ instance Parameterized Parameter where
...
+instance Parameterized Buffer where
+  flattenParameters _ = []
+  replaceOwnParameters = return
+
 instance Parameterized Double where
   flattenParameters _ = []
   replaceOwnParameters = return
 
-instance Parameterized [Int] where
+instance Parameterized Float where
```
Comment (tycho01) [inline hasktorch/src/Torch/NN.hs:75]: for this one I tried `instance (Scalar a) => Parameterized a`, but GHC felt that was not okay. :confused:

File: hasktorch/src/Torch/NN.hs [hunk 3]
```diff
@@ -153,7 +178,7 @@ data Conv2d =
   Conv2d { 
     conv2dWeight :: Parameter, 
     conv2dBias   :: Parameter
-    } deriving (Show, Generic)
+    } deriving (Show, Generic, Parameterized)
```
Comment (tycho01) [inline hasktorch/src/Torch/NN.hs:181]: we can do that now? that's great! :)

File: hasktorch/src/Torch/Tensor.hs [hunk 4]
```diff
@@ -310,6 +313,28 @@ instance {-# OVERLAPPING #-}TensorLike Bool where
...
+
+
+instance {-# OVERLAPPING #-}TensorLike a => TensorLike (a,a) where
+  asTensor' = error "Not implemented for tuple-type"
+  asTensor (a,b) = asTensor [a,b]
+  asValue v =
+    let [a,b] = asValue v
+    in (a,b)
+  _dtype = error "Not implemented for tuple-type"
+  _dims v = error "Not implemented for tuple-type"
```
Comment (tycho01) [inline hasktorch/src/Torch/Tensor.hs:334]: `[2]`?

File: hasktorch/src/Torch/Vision/Darknet/Forward.hs [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+      <*> pure stride
+      <*> pure layer_size
+      <*> pure (activation == "leaky")
+
+instance HasForward ConvolutionWithBatchNorm (Bool, Tensor) Tensor where
+  forward ConvolutionWithBatchNorm {..} (train, input) =
+    let pad = (layerSize - 1) `div` 2
+        activation = if isLeaky then flip I.leaky_relu 0.1 else id
+     in activation
+          $ batchNormForward batchNorm train 0.90000000000000002 1.0000000000000001e-05
```
Comment (tycho01) [inline hasktorch/src/Torch/Vision/Darknet/Forward.hs:66]: magic numbers :pouting_cat:

File: hasktorch/src/Torch/NN.hs (hunk 1 above)
Comment (junjihashimoto) [reply hasktorch/src/Torch/NN.hs]: Thanks for your suggestion.
I've changed the instance with generalized one.
Untyped instance can be fixed with 'OVERLAPS' pragma.
But typed instance can not be fixed with  'OVERLAPS' pragma.(I don't know why.)

**Comment by junjihashimoto** [comment]
@tycho01 Thank you for your reviewing.

File: hasktorch/src/Torch/Vision/Darknet/Forward.hs [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+  -- |  [1, 1, grid, grid]
+  Tensor
+gridX g = D.view [1, 1, g, g] $ D.repeat [g, 1] $ arange' (0 :: Int) g (1 :: Int)
+
+gridY ::
+  -- |  grid size
+  Int ->
+  -- |  [1, 1, grid, grid]
+  Tensor
+gridY g = D.contiguous $ D.view [1, 1, g, g] $ I.t $ D.repeat [g, 1] $ arange' (0 :: Int) g (1 :: Int)
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Vision/Darknet/Forward.hs:247]: https://stackoverflow.com/questions/49643225/whats-the-difference-between-reshape-and-view-in-pytorch
D.contiguous $ D.view  is the same as D.reshape

File: hasktorch/src/Torch/Vision/Darknet/Forward.hs [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+
+loadWeights :: Darknet -> String -> IO Darknet
+loadWeights (Darknet layers) weights_file = do
+  System.IO.withFile weights_file System.IO.ReadMode $ \handle -> do
+    _ <- BS.hGet handle (5 * 4) -- skip header
+    layers' <- forM layers $ \(i,layer) -> do
+      let cur_params = flattenParameters layer
+      new_params <- forM cur_params $ \param -> loadFloats handle (toDependent param) >>= makeIndependent
+      return $ (i,replaceParameters layer new_params)
+    return $ Darknet layers'
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Vision/Darknet/Forward.hs:484]: Conv2d/BatchNorm's paramaters are loaded incorrectly.

Comment (austinvhuang) [reply hasktorch/src/Torch/Vision/Darknet/Forward.hs:484]: Any idea why? I would like to try to do some work on loading pytorch models in the near future as well.

Comment (junjihashimoto) [reply hasktorch/src/Torch/Vision/Darknet/Forward.hs:484]: The data is stored in the weight file as follows.
1, Bias of batchnorm
2, Weight of batchnorm
3, RunningMean of batchnorm
4, RunningVar of batchnorm
5, Weight of Conv2d

Conv2d has no bias.
https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/master/models.py#L282-L317

The list of flattenParameters includes the bias of Conv2d with batchnorm.
Conv2d without batchnorm has bias.
Loader customization is needed.

Comment (junjihashimoto) [reply hasktorch/src/Torch/Vision/Darknet/Forward.hs:484]: It may be easier to fix the ADT.
